### PR TITLE
Parallelize masquerade lookup

### PR DESCRIPTION
**Pull request by myleshorton**

### CODE REVIEW

File: direct.go [hunk 1]
```diff
@@ -84,72 +84,65 @@ func (d *direct) providerFor(m *masquerade) *Provider {
...
-	log.Debug("Vetting one")
-	unvettedMasquerades := make([]*masquerade, 0, len(d.masquerades))
-	for _, m := range d.masquerades {
-		if m.lastSucceeded().IsZero() {
-			unvettedMasquerades = append(unvettedMasquerades, m)
+func (d *direct) findWorkingMasquerades() {
+	// vet masquerades in batches
+	const batchSize int = 25
+	var successful atomic.Uint32
+	for i := 0; i < len(d.masquerades) && successful.Load() < 4; i += batchSize {
```
Comment (reflog) [inline direct.go]: might be worth adding a constant

File: direct.go [hunk 2]
```diff
@@ -84,72 +84,65 @@ func (d *direct) providerFor(m *masquerade) *Provider {
...
+				if d.vetMasquerade(m) {
+					successful.Add(1)
+				}
+			}(d.masquerades[j])
 		}
+		wg.Wait()
 	}
+}
 
-	// Don't take more than 10 seconds to dial a masquerade for vetting
```
Comment (reflog) [inline direct.go:130]: i think this context timeout got lost in the refactor 🤔

File: direct.go [hunk 3]
```diff
@@ -551,3 +529,21 @@ func (ddf *directTransport) RoundTrip(req *http.Request) (resp *http.Response, e
...
+	r, err := http.NewRequest(req.Method, url.String(), body)
+	if err != nil {
+		return nil, err
+	}
+
+	for k, vs := range req.Header {
+		if !strings.EqualFold(k, "Host") {
+			v := make([]string, len(vs))
+			copy(v, vs)
+			r.Header[k] = v
```
Comment (reflog) [inline direct.go:491]: just out of curiosity, why not just direct assignment of vs? why the copy?

**Review by reflog — COMMENTED** [review]
LGTM, added a couple of small comments

**Comment by myleshorton** [comment]
Thanks @reflog! I'll look at these today, but I wonder if it might be worth just testing with this branch to see if it improves anything in your tests?

I'm also going to add much better tests of this to make sure it actually speeds things up just in tests.

**Comment by reflog** [comment]
> Thanks @reflog! I'll look at these today, but I wonder if it might be worth just testing with this branch to see if it improves anything in your tests?
> 
> I'm also going to add much better tests of this to make sure it actually speeds things up just in tests.

sounds good. made a branch in flashlight using this PR https://github.com/getlantern/flashlight/pull/1430
now going to deploy new pinger pointing to that branch...

File: direct.go (hunk 2 above)
Comment (myleshorton) [reply direct.go:130]: This code was...super confusing. Basically `dialWith` would dial all the theoretically unvetted masquerades (which would often be basically all of them) with a timeout of 10 seconds for the whole batch, but with a dial timeout of 10 seconds on each one. So, if one timed out dialing, we'd just be done.

The new code vets one masquerade at a time and just keeps the 10 second dial timeout, so I don't think we need the outer one.

**Comment by myleshorton** [comment]
OK I think this is close. I made things a bit more testable but definitely could add more tests.

File: cache.go [hunk 4]
```diff
@@ -87,6 +88,16 @@ func (d *direct) updateCache(cacheFile string) {
...
+		// Log the directory of the cache file and if it exists for debugging purposes
+		parent := filepath.Dir(cacheFile)
+		// check if the parent directory exists
+		if _, err := os.Stat(parent); err == nil {
+			// parent directory exists
+			log.Debugf("Parent directory of cache file exists: %v", parent)
+		} else {
+			// parent directory does not exist
+			log.Debugf("Parent directory of cache file does not exist: %v", parent)
+		}
```
Comment (myleshorton) [inline cache.go:100]: This is just for future debugging as the logs are showing issues here from time to time.

File: direct.go [hunk 5]
```diff
@@ -70,127 +68,91 @@ func (d *direct) loadCandidates(initial map[string]*Provider) {
...
+	var wg sync.WaitGroup
+	masqueradeSize := len(d.masquerades)
+	for j := start; j < start+batchSize && j < masqueradeSize; j++ {
+		wg.Add(1)
+		go func(m MasqueradeInterface) {
+			defer wg.Done()
+			if d.vetMasquerade(m) {
+				successful.Add(1)
+			}
+		}(d.masquerades[j])
```
Comment (myleshorton) [inline direct.go:127]: The above is the most important change along with cleaning up the 10 second timeout thing

File: fronted.go [hunk 6]
```diff
@@ -70,141 +105,104 @@ func (d *direct) loadCandidates(initial map[string]*Provider) {
...
-		if m.lastSucceeded().IsZero() {
-			unvettedMasquerades = append(unvettedMasquerades, m)
-		}
+func (f *fronted) vetBatch(start, batchSize int, successful *atomic.Uint32) {
+	log.Debugf("Vetting masquerade batch %d-%d", start, start+batchSize)
+	var wg sync.WaitGroup
+	masqueradeSize := len(f.masquerades)
+	for j := start; j < start+batchSize && j < masqueradeSize; j++ {
+		wg.Add(1)
+		go func(m MasqueradeInterface) {
```
Comment (myleshorton) [inline fronted.go:159]: This is the most important change -- just parallelizing the vetting process.

**Comment by myleshorton** [comment]
OK I ended up cleaning up the naming a fair bit -- "direct" was a relic of my first pass at this circa 2015 or something when we thought of this as a unique implementation of domain fronting that uses IPs directly versus resolving the DNS. That is actually somewhat unique, but the name was just confusing.

**Comment by myleshorton** [comment]
What do you think about pulling this in @reflog, perhaps after another round of using the latest with pinger?

**Comment by myleshorton** [comment]
oh hmnn will fix tests!

**Comment by reflog** [comment]
pinger updated

**Comment by myleshorton** [comment]
OK I'm pulling this in! Thanks @WendelHime!

**Pull request merged**